src/Common/src/System/Net/Http/TlsCertificateExtensions.cs

```diff
@@ -0,0 +1,129 @@
+                // But do this only if needed. No point going over it, if we already have established that our cert has the
+                // required extension.
+                if (!foundEku)
+                {
+                    X509EnhancedKeyUsageExtension enhancedUsageExt = extension as X509EnhancedKeyUsageExtension;
+                    if (enhancedUsageExt != null)
+                    {
+                        foundEku = true;
+                        isClientAuth = false;
+                        foreach(Oid oid in enhancedUsageExt.EnhancedKeyUsages)
+                        {
+                            if (string.Equals(ClientCertificateOid, oid.Value))
+                            {
+                                isClientAuth = true;
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                /// Check if the extension is a key usage extension.
+                /// No point going over it if we have already established that out cert has digital signature
+                if (!foundKeyUsages)
+                {
+                    X509KeyUsageExtension usageExt = extension as X509KeyUsageExtension;
+                    if (usageExt != null)
+                    {
+                        foundKeyUsages = true;
+                        isDigitalSignature = (usageExt.KeyUsages & X509KeyUsageFlags.DigitalSignature) != 0;
+                    }
+                }
+
+                if (foundKeyUsages && foundEku)
+                {
+                    break;
+                }
+            }
+
+            return isClientAuth && isDigitalSignature;
+        }
+
+        /// <summary>
+        ///   Returns a new collection containing valid client certificates from the given X509Certificate2Collection
+        /// </summary>
+        internal static bool TryFindClientCertificate(this X509Certificate2Collection certificates,
+                                                      ISet<string> allowedIssuers,
+                                                      out X509Certificate2 clientCertificate,
+                                                      out X509Chain clientCertChain)
+        {
+            clientCertificate = null;
+            clientCertChain = null;
+            if (certificates == null)
+            {
+                return false;
+            }
+
+            DateTime now = DateTime.Now;
+            foreach(X509Certificate2 cert in certificates)
+            {
+                if (cert.HasPrivateKey && now > cert.NotBefore && now < cert.NotAfter)
+                {
+                    if (IsClientCertificate(cert))
+                    {
+                        X509Chain chain = new X509Chain();
+                        //TODO consider changing the following to AllFlags
+                        chain.ChainPolicy.VerificationFlags = X509VerificationFlags.NoFlag;
```

Comment:
AllowUnknownCertificateAuthority is probably something you want to set, because otherwise it requires that the client trust their own cert, which isn't always true.